source/midi_stats.py

```diff
@@ -0,0 +1,500 @@
+    else:
+        return None
+
+
+def number_of_notes_per_second(pm):
+    if pm.instruments:
+        return (
+            len([note for instrument in pm.instruments for note in instrument.notes])
+            / pm.get_end_time()
+        )
+    else:
+        return None
+
+
+def shortest_note_length(pm):
+    if pm.instruments:
+        return min(
+            [
+                note.end - note.start
+                for instrument in pm.instruments
+                for note in instrument.notes
+            ]
+        )
+    else:
+        return None
+
+
+def longest_note_length(pm):
+    if pm.instruments:
+        return max(
+            [
+                note.end - note.start
+                for instrument in pm.instruments
+                for note in instrument.notes
+            ]
+        )
+    else:
+        return None
+
+
+def main_key_signature(pm):
+    if pm.key_signature_changes:
+        return pm.key_signature_changes[0].key_number
+    else:
+        return None
+
+
+def n_key_changes(pm):
+    if pm.key_signature_changes:
+        return len(pm.key_signature_changes)
+    else:
+        return None
+
+
+def n_tempo_changes(pm):
+    return len(pm.get_tempo_changes())
+
+
+def average_tempo(pm):
+    try:
+        return round(pm.estimate_tempo())
+    except Exception:
+        return None
+
+
+def tempo_changes(pm):
+    return [[pm.get_tempo_changes()]]
+
+
+def main_time_signature(pm):
+    if pm.time_signature_changes:
+        return [
+            str(ts.numerator) + "/" + str(ts.denominator)
+            for ts in pm.time_signature_changes
+        ][0]
+    else:
+        return None
+
+
+def n_time_signature_changes(pm):
+    if pm.time_signature_changes:
+        return len(pm.time_signature_changes)
+    else:
+        return None
+
+
+def all_time_signatures(pm):
+    if pm.time_signature_changes:
+        return [
+            [
+                str(ts.numerator) + "/" + str(ts.denominator)
+                for ts in pm.time_signature_changes
+            ]
+        ]
+    else:
+        return None
+
+
+def four_to_the_floor(pm):
+    if pm.time_signature_changes:
+        time_signatures = [
+            str(ts.numerator) + "/" + str(ts.denominator)
+            for ts in pm.time_signature_changes
+        ]
+        # check if time_signatures contains exclusively '2/4' or '4/4'
+        return (
+            all([ts == "4/4" for ts in time_signatures]) and len(time_signatures) == 1
+        )
+    else:
+        return None
+
+
+def track_length_in_seconds(pm):
+    return pm.get_end_time()
+
+
+def lyrics_number_of_words(pm):
+    if pm.lyrics:
+        return len([l.text for l in pm.lyrics])
+    else:
+        return None
+
+
+def lyrics_number_of_unique_words(pm):
+    if pm.lyrics:
+        return len(set([l.text for l in pm.lyrics]))
+    else:
+        return None
+
+
+def lyrics_boolean(pm):
+    if pm.lyrics:
+        return True
+    else:
+        return False
+
+
+class MidiStats:
```

Comment:
For me it would make more sense if the function above were in the class? But maybe you had a reason not to put them in there.

Reply:
Yes it does make sense!